src/core/manager.c

```diff
@@ -556,6 +558,46 @@ static int manager_default_environment(Manager *m) {
         return 0;
 }
 
+static const struct {
+        ExecDirectoryType directory;
+        uint64_t type;
+        const char *suffix;
+} paths_system[_EXEC_DIRECTORY_MAX] = {
+        { EXEC_DIRECTORY_RUNTIME, SD_PATH_SYSTEM_RUNTIME, NULL },
+        { EXEC_DIRECTORY_STATE, SD_PATH_SYSTEM_STATE_PRIVATE, NULL },
+        { EXEC_DIRECTORY_CACHE, SD_PATH_SYSTEM_STATE_CACHE, NULL },
+        { EXEC_DIRECTORY_LOGS, SD_PATH_SYSTEM_STATE_LOGS, NULL },
+        { EXEC_DIRECTORY_CONF, SD_PATH_SYSTEM_CONFIGURATION, NULL },
+}, paths_user[_EXEC_DIRECTORY_MAX] = {
+        { EXEC_DIRECTORY_RUNTIME, SD_PATH_USER_RUNTIME, NULL },
+        { EXEC_DIRECTORY_STATE, SD_PATH_USER_CONFIGURATION, NULL },
+        { EXEC_DIRECTORY_CACHE, SD_PATH_SYSTEM_STATE_CACHE, NULL },
+        { EXEC_DIRECTORY_LOGS, SD_PATH_SYSTEM_CONFIGURATION, "log" },
+        { EXEC_DIRECTORY_CONF, SD_PATH_SYSTEM_CONFIGURATION, NULL },
+};
+
+static int manager_setup_prefix(Manager *m) {
+        int i, r;
+
+        assert(m);
+
+        if (MANAGER_IS_SYSTEM(m))
+                for (i = 0; i < _EXEC_DIRECTORY_MAX; i++) {
+                        r = sd_path_home(paths_system[i].type, paths_system[i].suffix, &m->prefix[paths_system[i].directory]);
+                        if (r < 0)
+                                return r;
+                }
+        else
+                for (i = 0; i < _EXEC_DIRECTORY_MAX; i++) {
+                        r = sd_path_home(paths_user[i].type, paths_user[i].suffix, &m->prefix[paths_user[i].directory]);
+                        if (r < 0)
+                                return r;
+                }
```

Comment:
I'd first pick the array to use, assign that to a local variable, and then operate on that local variable, so that we only need a single for() loop. Yes, this would require defining a new type for the struct you define above, but if you do all that as private type inside the function you'd not even litter the namespace for this...